tests/unit/s2n_examples_test.c

```diff
@@ -0,0 +1,329 @@
+
+    DEFER_CLEANUP(struct s2n_blob input = { 0 }, s2n_free);
+    RESULT_GUARD_POSIX(s2n_alloc(&input, S2N_TEST_BYTES_TO_SEND));
+    RESULT_GUARD(s2n_get_public_random_data(&input));
+
+    pid_t client_pid = fork();
+    if (client_pid == 0) {
+        /* Suppress stdout when running the examples.
+         * This only affects the new client process.
+         */
+        fclose(stdout);
+
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+        EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client, &io_pair));
+
+        EXPECT_OK(scenario_fn(client, &input));
+
+        exit(EXIT_SUCCESS);
+    }
+
+    pid_t server_pid = fork();
+    if (server_pid == 0) {
+        /* Suppress stdout when running the examples.
+         * This only affects the new server process.
+         */
+        fclose(stdout);
+
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+
+        EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server, &io_pair));
+
+        EXPECT_OK(scenario_fn(server, &input));
+
+        exit(EXIT_SUCCESS);
+    }
+
+    int status = 0;
+    RESULT_ENSURE_EQ(waitpid(client_pid, &status, 0), client_pid);
+    RESULT_ENSURE_EQ(status, EXIT_SUCCESS);
+    RESULT_ENSURE_EQ(waitpid(server_pid, &status, 0), server_pid);
+    RESULT_ENSURE_EQ(status, EXIT_SUCCESS);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_run_failure_tests()
+{
+    uint8_t buffer[100] = { 0 };
+    size_t buffer_size = sizeof(buffer);
+
+    DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+            s2n_connection_ptr_free);
+    EXPECT_SUCCESS(s2n_connection_set_blinding(conn, S2N_SELF_SERVICE_BLINDING));
+
+    pid_t pid = fork();
+    if (pid == 0) {
+        /* Suppress stdout AND stderr when running the examples.
+         * This only affects the new process.
+         */
+        fclose(stdout);
+        fclose(stderr);
+
+        EXPECT_EQUAL(s2n_negotiate_basic_example(conn), S2N_FAILURE);
+        EXPECT_EQUAL(s2n_send_basic_example(conn, buffer, buffer_size), S2N_FAILURE);
+        EXPECT_EQUAL(s2n_sendv_basic_example(conn, buffer, buffer_size), S2N_FAILURE);
+        EXPECT_EQUAL(s2n_recv_basic_example(conn, buffer, buffer_size), S2N_FAILURE);
+        EXPECT_EQUAL(s2n_recv_echo_example(conn, buffer, buffer_size), S2N_FAILURE);
+
+        exit(EXIT_SUCCESS);
+    }
+
+    int status = 0;
+    RESULT_ENSURE_EQ(waitpid(pid, &status, 0), pid);
+    RESULT_ENSURE_EQ(status, EXIT_SUCCESS);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
```

Comment:
Is the main purpose of this file to test the examples and make sure they're correct? Or is it supposed to be read by users? 

In general, I worry here that we're taking stuff out of the usage guide and turning it into unit tests, and we're losing readability in the process. Like we don't want our usage guide to just be "go look at our unit tests", which are not readable to users at all.

Reply:
The main purpose of this file is to test the examples and make sure they're correct. Customers are absolutely not supposed to read this file.